### ISSUE TITLE
Consider using react-native-quick-crypto

### ISSUE THREAD

**Issue by mrousavy** [BlueWallet/BlueWallet#5086]
Hey!

I noticed this repo uses **react-native-crypto**, which itself uses **crypto-browserify** under the hood. **crypto-browserify** is a JS-based implementation for crypto, which is slower than Margelo's [**react-native-quick-crypto**](https://github.com/margelo/react-native-quick-crypto), since that uses C++ (OpenSSL) with JSI under the hood.

If you want to replace **react-native-crypto** with the faster **react-native-quick-crypto**, I'm happy to help out or answer questions! 😄

**Comment by Overtorment** [comment]
interesting! is that future-proof for RN70+ ?

**Comment by mrousavy** [comment]
Yep, we'll make sure to make it compatible with RN 0.70 alongside my other libraries (MMKV, VisionCamera)

**Comment by Overtorment** [comment]
sounds good!
seeing that `react-native-crypto` is  being deprecated (as stated in their readme), help with migration would be greatly appreciated!

**Comment by Overtorment** [comment]
@mrousavy is there a minimum RN version required? getting some build errors here:  https://github.com/BlueWallet/BlueWallet/pull/5255

**Comment by Overtorment** [comment]
weird, but builds (& e2e) succeded on CI, but i get errors when building android locally:
[build.log](https://github.com/BlueWallet/BlueWallet/files/10329599/build.log)

**Comment by Overtorment** [comment]
```
* What went wrong:
Execution failed for task ':app:mergeReleaseNativeLibs'.
> A failure occurred while executing com.android.build.gradle.internal.tasks.MergeJavaResWorkAction
   > 2 files found with path 'lib/x86_64/libreact_nativemodule_core.so' from inputs:
      - /home/ghost/Documents/BlueWallet/node_modules/react-native-quick-crypto/android/build/intermediates/library_jni/release/jni
      - /home/ghost/.gradle/caches/transforms-3/2b34732fc3cc3ee0f677949ea43e2d53/transformed/jetified-react-native-0.67.5/jni
     If you are using jniLibs and CMake IMPORTED targets, see
     https://developer.android.com/r/tools/jniLibs-vs-imported-targets

```

**Comment by Overtorment** [comment]
also getting unit test failures:

```

 FAIL  tests/unit/lightning-ldk-wallet.test.js
  ● Test suite failed to run

    Failed to install react-native-quick-crypto: The native `QuickCrypto` Module could not be found.
    * Make sure react-native-quick-crypto is correctly autolinked (run `npx react-native config` to verify)
    * Make sure you ran `pod install` in the ios/ directory.
    * Make sure you rebuilt the app.


```

**Comment by Overtorment** [comment]
ok, all issues sorted.
i had to disable `react-native-quick-crypto` for nodejs environment (to run jest unit tests which rely on plain old node's crypto).
also added a weird workaround in build gradle so it picks first binary out of multiple. dont know why but it reproduces only on my machine and not con CI.

todo: do performance tests!

**Comment by mrousavy** [comment]
oh yeah, for nodejs the native JSI stuff won't work. let me know about your perf findings

**Comment by Overtorment** [comment]
a bit weird, but im observing ~30% performance degradation with this package.

testing code that im using: https://github.com/BlueWallet/BlueWallet/pull/5257


```js
          const secret = 'abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon about';
          const w = new HDSegwitBech32Wallet();
          w.setSecret(secret);

          const start = Date.now();
          let num;
          for (num = 0; num < 1000; num++) {
            w._getExternalAddressByIndex(num);
            if (Date.now() - start > 10 * 1000) {
              break;
            }
          }

```


where `_getExternalAddressByIndex`  boils down to

```js
      const xpub = this._zpubToXpub(this.getXpub());
      const hdNode = bip32.fromBase58(xpub);
      this._node0 = hdNode.derive(node);

// ...

    return bitcoin.payments.p2wpkh({
      pubkey: hdNode.publicKey,
    }).address;

```


PS. csprng is not accessed here

**Comment by Overtorment** [comment]
^^^ on android theres a degradation, on ios there is no difference

**Comment by Overtorment** [comment]
@mrousavy thoughts?

**Comment by Overtorment** [comment]
@junderw any insights on this..?

**Comment by junderw** [comment]
This is something we've noticed with WASM too.

Whenever there's a communication between the JS runtime and an outside runtime JSI/WASM etc. data is usually serialized and deserialized both ways.

We got that down with WASM by avoiding one-size-fits-all libraries like wasm-pack and creating our own near-zero-cost serialization layer that isn't really serializing, it uses the data formats exactly as they are used in C/Rust and on the JS side just copying Buffers to UInt8Arrays.

Considering the derivation code is relatively a small portion of the logical steps in each loop, the latency between the native portion and the JS portion is probably causing the issues...

Not sure though.

**Comment by Overtorment** [comment]
also, we used `react-native-crypto` before, and its not a pure-js implementation. its crypto stuff implemented in native.
maybe its just more performant than `react-native-quick-crypto`...

**Comment by mrousavy** [comment]
`react-native-crypto` is **not native**. It's a browser polyfill (so JS based solution), and only the `getRandomValues()` func is a native function.

`react-native-quick-crypto` is truly native. It's really weird that this is slower.... 🤔

**Comment by mrousavy** [comment]
The bottleneck was `pbkdf2`, which is a key derivation function that was insanely slow in JS. Both on Hermes and JSC. With react-native-quick-crypto (C++), it was waaaay faster.

**Comment by Overtorment** [comment]
aah i think i remember this one, i implemented it and `scrypt` in native code a few years back for RN to speed things up for BW
( https://github.com/BlueWallet/react-native-blue-crypto/blob/e9b60605dbdb24de720e923f740912a4930383ff/android/src/main/java/com/reactlibrary/SCrypt.java#L216 )


@junderw where do we utilise `pbkdf2` in `bitcoinjs` suite so i can better test performance gains? apart from BIP38

**Comment by mrousavy** [comment]
Just to confirm - did you test this in Release mode?

In debug, JSI has lots of overhead, and there's on RN version where JSI is actually really slow (I think 0.70?) - so make sure to test it either on RN 0.68 or RN 0.71 (maybe we need some changes in RN QuickCrypto so it builds for RN 0.71).

**Comment by Overtorment** [comment]
i think i tested it on ios release build (apparently without hermes), with 0.67.5

updating to 0.71 is on our todo list

**Comment by mrousavy** [comment]
> i think i tested it on ios release build (apparently without hermes), with 0.67.5

That should be fine. hmm

**Comment by junderw** [comment]
pbkdf2 is mainly used in the bip39 module for converting a mnemonic string into the seed Buffer for use with bip32 HD wallets... bip32 mainly uses HMAC-SHA512

**Comment by otech47** [comment]
> i had to disable react-native-quick-crypto for nodejs environment (to run jest unit tests which rely on plain old node's crypto).


running into this problem as well, how did you configure this exactly?

**Comment by Overtorment** [comment]
> > i had to disable react-native-quick-crypto for nodejs environment (to run jest unit tests which rely on plain old node's crypto).
> 
> running into this problem as well, how did you configure this exactly?

take a look at PR

**Comment by keyberth-rengel** [comment]
> https://github.com/BlueWallet/react-native-blue-crypto/blob/e9b60605dbdb24de720e923f740912a4930383ff/ android/src/main/java/com/reactlibrary/SCrypt.java#L216

Tengo este mismo problema ¿has encontrado la forma de solucionarlo ?
